pkg/app/piped/eventwatcher/eventwatcher.go

```diff
@@ -184,23 +180,58 @@ func (w *watcher) run(ctx context.Context, repo git.Repo, repoCfg config.PipedRe
 				}
 				continue
 			}
+			// Check whether the config file exists in .pipe/ or not and updates values if it exists.
+			// NOTE: This will be removed.
 			cfg, err := config.LoadEventWatcher(repo.GetPath(), includedCfgs, excludedCfgs)
-			if errors.Is(err, config.ErrNotFound) {
-				w.logger.Info("configuration file for Event Watcher not found",
+			if !errors.Is(err, config.ErrNotFound) && err != nil {
+				w.logger.Error("failed to load configuration file for Event Watcher",
 					zap.String("repo-id", repoCfg.RepoID),
 					zap.Error(err),
 				)
 				continue
 			}
-			if err != nil {
-				w.logger.Error("failed to load configuration file for Event Watcher",
+			if errors.Is(err, config.ErrNotFound) {
+				w.logger.Info("configuration file for Event Watcher in .pipe/ not found",
 					zap.String("repo-id", repoCfg.RepoID),
 					zap.Error(err),
 				)
+			} else {
+				if err := w.updateValues(ctx, repo, repoCfg.RepoID, cfg.Events, commitMsg); err != nil {
+					w.logger.Error("failed to update the values",
+						zap.String("repo-id", repoCfg.RepoID),
+						zap.Error(err),
+					)
+				}
+			}
+			// If event watcher config exist in the application config file, they are handled.
+			resp, err := w.apiClient.ListApplications(ctx, &pipedservice.ListApplicationsRequest{})
+			if err != nil {
+				w.logger.Error("failed to list registered application", zap.Error(err))
 				continue
 			}
-			if err := w.updateValues(ctx, repo, repoCfg.RepoID, cfg.Events, commitMsg); err != nil {
-				w.logger.Error("failed to update the values",
+			cfgs := make([]config.EventWatcherConfig, 0, len(resp.Applications))
+			for _, app := range resp.Applications {
+				if app.GitPath.Repo.Id != repoCfg.RepoID {
+					continue
+				}
+				appCfg, err := config.LoadApplication(repo.GetPath(), app.GitPath.GetApplicationConfigFilePath(), app.Kind)
+				if err != nil {
+					w.logger.Error("failed to load application configuration", zap.Error(err))
+					continue
+				}
+				if appCfg.EventWatcher == nil {
+					continue
+				}
+				cfgs = append(cfgs, appCfg.EventWatcher...)
+			}
+			if len(cfgs) == 0 {
+				w.logger.Info("configuration for Event Watcher in application configuration not found",
+					zap.String("repo-id", repoCfg.RepoID),
+				)
+				continue
```

Comment:
How about having a cache layer to avoid recalculating for unchanged applications?

Reply:
Nice catch!
What do you this about like this? https://github.com/pipe-cd/pipecd/pull/3729/commits/58c4aa8085dd3ac694951985b9c0ff5f67dccd6c

Reply:
Thank you. Let me review the changes.